src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/SalesforceBaseSourceConfig.java

```diff
@@ -42,55 +50,70 @@
 
-  @Name(SalesforceSourceConstants.PROPERTY_DATETIME_FILTER)
-  @Description("Salesforce SObject query datetime filter. Example: 2019-03-12T11:29:52Z, LAST_WEEK")
+  @Name(SalesforceSourceConstants.PROPERTY_DATETIME_AFTER)
+  @Description("Salesforce SObject query datetime filter. Example: 2019-03-12T11:29:52Z")
   @Nullable
   @Macro
-  private String datetimeFilter;
+  private String datetimeAfter;
+
+  @Name(SalesforceSourceConstants.PROPERTY_DATETIME_BEFORE)
+  @Description("Salesforce SObject query datetime filter. Example: 2019-03-12T11:29:52Z")
+  @Nullable
+  @Macro
+  private String datetimeBefore;
 
   @Name(SalesforceSourceConstants.PROPERTY_DURATION)
   @Description("Salesforce SObject query duration. Time unit: hours. Default value: 0")
   @Nullable
   @Macro
-  private Integer duration;
+  private String duration;
 
   @Name(SalesforceSourceConstants.PROPERTY_OFFSET)
   @Description("Salesforce SObject query offset. Time unit: hours. Default value: 0")
   @Nullable
   @Macro
-  private Integer offset;
+  private String offset;
 
   protected SalesforceBaseSourceConfig(String referenceName,
                                        String consumerKey,
                                        String consumerSecret,
                                        String username,
                                        String password,
                                        String loginUrl,
-                                       @Nullable String datetimeFilter,
-                                       @Nullable Integer duration,
-                                       @Nullable Integer offset) {
+                                       @Nullable String datetimeAfter,
+                                       @Nullable String datetimeBefore,
+                                       @Nullable String duration,
+                                       @Nullable String offset) {
     super(referenceName, consumerKey, consumerSecret, username, password, loginUrl);
-    this.datetimeFilter = datetimeFilter;
+    this.datetimeAfter = datetimeAfter;
+    this.datetimeBefore = datetimeBefore;
     this.duration = duration;
     this.offset = offset;
   }
 
-  public int getDuration() {
-    return Objects.isNull(duration) ? SalesforceSourceConstants.DURATION_DEFAULT : duration;
+  public Map<ChronoUnit, Integer> getDuration() {
+    return Strings.isNullOrEmpty(duration) ? Collections.emptyMap() : extractRangeValue(duration);
   }
 
-  public int getOffset() {
-    return Objects.isNull(offset) ? SalesforceSourceConstants.OFFSET_DEFAULT : offset;
+  public Map<ChronoUnit, Integer> getOffset() {
+    return Strings.isNullOrEmpty(offset) ? Collections.emptyMap() : extractRangeValue(offset);
   }
 
   @Nullable
-  public String getDatetimeFilter() {
-    return datetimeFilter;
+  public String getDatetimeAfter() {
+    return datetimeAfter;
+  }
+
+  @Nullable
+  public String getDatetimeBefore() {
```

Comment:
it's better if these return ZonedDateTime, so that anything that calls it is guaranteed to use valid values.

Reply:
that means the correctness of this class is coupled with the correctness of an unrelated class. If somebody uses it somewhere else in the future, they might not validate the input, resulting in buggy code. It is better to have APIs where correctness is guaranteed if possible.